src/main/java/si/mazi/rescu/clients/ApacheConnection.java

```diff
@@ -0,0 +1,228 @@
+import org.apache.http.client.methods.HttpDelete;
+import org.apache.http.client.methods.HttpEntityEnclosingRequestBase;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpHead;
+import org.apache.http.client.methods.HttpOptions;
+import org.apache.http.client.methods.HttpPatch;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpPut;
+import org.apache.http.client.methods.HttpRequestBase;
+import org.apache.http.entity.ByteArrayEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import si.mazi.rescu.HttpMethod;
+
+public class ApacheConnection implements HttpConnection {
+    
+    private static final Logger log = LoggerFactory.getLogger(ApacheConnection.class);
+    
+    private final String url;
+    private final Proxy proxy;
+    private final HttpClientBuilder builder = HttpClients.custom();
+    private final RequestConfig.Builder requestConfig = RequestConfig.custom();
+            
+    private boolean executed = false;
+    private CloseableHttpResponse res;
+    private HttpMethod method = HttpMethod.GET;
+    private ByteArrayOutputStream out;
+    private final Map<String, String> headers = new HashMap<>();
+    
+    private CloseableHttpResponse response;
+
+    private ApacheConnection(String url, Proxy proxy) {
+        super();
+        this.url = url;
+        this.proxy = proxy;
+    }
+    
+    public static HttpConnection create(String url, Proxy proxy) throws MalformedURLException, IOException {
+        return new ApacheConnection(url, proxy);
+    }
+    
+    private void exec() {
+        if (executed) {
+            return;
+        }
+        HttpClientBuilder clientBuilder = builder
+                .setDefaultRequestConfig(requestConfig.build());
+        if (proxy != null && proxy != Proxy.NO_PROXY) {
+            InetSocketAddress address = (InetSocketAddress) proxy.address(); 
+            HttpHost proxy = new HttpHost(address.getHostName(), address.getPort());
+            clientBuilder.setProxy(proxy);
+        }
+        
+        CloseableHttpClient client = clientBuilder
+                .build();
+        
+        HttpRequestBase request = createRequest(method, url);
+        
+        if (request instanceof HttpEntityEnclosingRequestBase && out != null) {
+            HttpEntityEnclosingRequestBase req = (HttpEntityEnclosingRequestBase) request;
+            HttpEntity entity = req.getEntity();
+            if (entity == null) {
+                entity = new ByteArrayEntity(out.toByteArray());
+                req.setEntity(entity);
+            }
+        }
+        headers.forEach(request::addHeader);
+        try {
+            response = client.execute(request);
+            
+        } catch (Throwable e) {
```

Comment:
Handle this better: IOException should be declared, non-checked exceptions should not be wrapped etc.